File: 04_AJIN/backend/schemas/equipment.py
"""Day 6 Phase 1 — 설비/공정 AI Pydantic 스키마.

Module F (Equipment & Process AI) 의 12 엔드포인트 요청/응답 모델.
features/equipment/* 19 모듈을 React UI 로 노출하기 위한 wire 포맷.
"""

from __future__ import annotations

from typing import Literal, Optional, Any

from pydantic import BaseModel, Field


# ═══════════════════════════════════════════════════════════
# 1. 공통 — Severity / Risk Level
# ═══════════════════════════════════════════════════════════

Severity = Literal["critical", "warning", "info"]
RiskLevel = Literal["critical", "warning", "normal"]
ProcessStatus = Literal["good", "warning", "critical"]


# ═══════════════════════════════════════════════════════════
# 2. Dashboard Overview — GET /equipment/dashboard/overview
# ═══════════════════════════════════════════════════════════


class ProcessHealthCard(BaseModel):
    """5공정 건강 카드 (CCH/OBC/범퍼빔/도어/볼시트)."""

    process_id: str
    process_name: str
    status: ProcessStatus
    current_cpk: float
    cpk_trend: str = "stable"
    violation_count: int = 0
    violated_rules: list[int] = Field(default_factory=list)
    risk_level: RiskLevel = "normal"
    anomaly_rate: float = 0.0


class EquipmentTypeCard(BaseModel):
    """7장비 카드 (프레스/용접기/로봇/사출기/CNC/레이저/공통설비)."""

    type: str
    icon: str
    codes: int
    key_metric: str
    color: str


class DashboardMetrics(BaseModel):
    """대시보드 5종 핵심 메트릭."""

    error_codes_total: int = 0
    error_codes_critical: int = 0
    molds_total: int = 0
    molds_warning: int = 0
    molds_critical: int = 0
    spc_processes: int = 0
    inspections_templates: int = 0
    inspections_recent: int = 0


class MLAlert(BaseModel):
    """ML 경고 카드."""

    level: str
    source: str
    message: str


class OverviewResponse(BaseModel):
    """GET /equipment/dashboard/overview 응답."""

    processes: list[ProcessHealthCard]
    equipment_types: list[EquipmentTypeCard]
    metrics: DashboardMetrics
    ml_alerts: list[MLAlert] = Field(default_factory=list)


# ═══════════════════════════════════════════════════════════
# 3. SPC Chart — GET /equipment/spc/{process_id}
# ═══════════════════════════════════════════════════════════


class NelsonViolationItem(BaseModel):
    """Nelson 8 Rules 위반 항목."""

    rule_number: int
    rule_name: str
    description: str
    severity: Severity
    points: list[int] = Field(default_factory=list)
    recommended_action: str = ""
    chart_annotation: str = ""


class SPCData(BaseModel):
    """SPC 관리도 데이터 (Plotly 시각화용)."""

    process_id: str
    process_name: str
    timestamps: list[int]
    values: list[float]
    mean: float
    sigma: float
    ucl: float
    lcl: float
    sigma_1_upper: float
    sigma_1_lower: float
    sigma_2_upper: float
    sigma_2_lower: float
    usl: Optional[float] = None
    lsl: Optional[float] = None


class SPCResponse(BaseModel):
    """GET /equipment/spc/{process_id} 응답."""

    data: SPCData
    violations: list[NelsonViolationItem]
    out_of_control: bool = False
    violation_count: int = 0


# ═══════════════════════════════════════════════════════════
# 4. SPC Recent Violations — GET /equipment/spc/violations/recent
# ═══════════════════════════════════════════════════════════


class RecentViolation(BaseModel):
    """최근 SPC 위반 — RTDB push 대상."""

    id: str
    process_id: str
    process_name: str
    rule_number: int
    severity: Severity
    message: str
    timestamp: int  # ms epoch


class ViolationsResponse(BaseModel):
    items: list[RecentViolation]
    total: int


# ═══════════════════════════════════════════════════════════
# 5. ML Error Search — POST /equipment/error/search
# ═══════════════════════════════════════════════════════════


class ErrorSearchRequest(BaseModel):
    query: str
    top_k: int = 5
    equipment_filter: Optional[str] = None


class ErrorSearchResult(BaseModel):
    code: str
    equipment_type: str
    category: str
    description: str
    cause: str
    action: str
    severity: str
    score: float
    rank: int


class CausalityInfo(BaseModel):
    causes: list[str] = Field(default_factory=list)
    actions: list[str] = Field(default_factory=list)


class ManualExcerpt(BaseModel):
    content: str
    source: str = ""
    page: str = ""
    relevance: float = 0.0


class ErrorSearchResponse(BaseModel):
    results: list[ErrorSearchResult]
    causality: Optional[CausalityInfo] = None
    manual_excerpts: list[ManualExcerpt] = Field(default_factory=list)


# ═══════════════════════════════════════════════════════════
# 6. Error Categories — GET /equipment/error/categories
# ═══════════════════════════════════════════════════════════


class CategoryGroup(BaseModel):
    equipment_type: str
    symptoms: list[str]


class ErrorCategoriesResponse(BaseModel):
    groups: list[CategoryGroup]
    total_synonyms: int = 0


# ═══════════════════════════════════════════════════════════
# 7. Markov Chain — GET /equipment/markov/{error_code}
# ═══════════════════════════════════════════════════════════


class MarkovPrediction(BaseModel):
    code: str
    category: str
    equipment_type: str
    probability: float
    expected_delay_hours: float
    description: str
    recommended_action: str


class CascadeStep(BaseModel):
    code: str
    category: str
    probability: float
    expected_delay_hours: float


class CascadeChainItem(BaseModel):
    steps: list[CascadeStep]
    total_probability: float
    total_hours: float


class MarkovResponse(BaseModel):
    current_code: str
    current_category: str
    next_predictions: list[MarkovPrediction]
    cascade_chains: list[CascadeChainItem] = Field(default_factory=list)
    risk_level: RiskLevel = "normal"
    prevention_message: str = ""


# ═══════════════════════════════════════════════════════════
# 8. Mold list + XGBoost — GET /equipment/molds
# ═══════════════════════════════════════════════════════════


class MoldItem(BaseModel):
    mold_id: str
    mold_name: str
    mold_type: str = ""
    part_name: str = ""
    current_shots: int = 0
    max_shots: int = 0
    life_percent: float = 0.0
    remaining_shots: int = 0
    status: str = "active"

    # XGBoost 예측 (선택 — 모델 파일 부재 시 None)
    predicted_remaining_life: Optional[int] = None
    predicted_replacement_date: Optional[str] = None
    risk_level: Optional[RiskLevel] = None
    confidence_interval: Optional[list[int]] = None


class MoldsResponse(BaseModel):
    items: list[MoldItem]
    total: int
    critical: int = 0
    warning: int = 0
    active: int = 0


# ═══════════════════════════════════════════════════════════
# 9. MTBF — GET /equipment/mtbf
# ═══════════════════════════════════════════════════════════


class MTBFItem(BaseModel):
    machine_id: str
    machine_name: str
    total_repairs: int
    mtbf_days: float
    mtbf_std_days: float = 0.0
    last_repair_date: str = ""
    next_predicted_date: str = ""
    days_until_next: int = 0
    risk_level: str = "정상"
    avg_repair_hours: float = 0.0
    avg_repair_cost: float = 0.0
    seasonal_pattern: dict[str, float] = Field(default_factory=dict)


class MTBFTopCost(BaseModel):
    machine_name: str
    total_cost: float


class MTBFResponse(BaseModel):
    items: list[MTBFItem]
    top5_cost: list[MTBFTopCost] = Field(default_factory=list)
    seasonal_message: str = ""
    machines_attention: int = 0


# ═══════════════════════════════════════════════════════════
# 10. ML Engines Status — GET /equipment/ml-engines/status
# ═══════════════════════════════════════════════════════════


EngineStatus = Literal["online", "offline", "warning"]


class MLEngineStatus(BaseModel):
    """7종 ML 모델 개별 상태."""

    id: str
    name_en: str
    name_ko: str
    library: str
    status: EngineStatus
    accuracy: Optional[float] = None
    last_trained: Optional[str] = None
    description: str = ""


class MLEnginesStatusResponse(BaseModel):
    engines: list[MLEngineStatus]
    online_count: int = 0
    total: int = 7


# ═══════════════════════════════════════════════════════════
# 11. Manual RAG — POST /equipment/manual/search
# ═══════════════════════════════════════════════════════════


class ManualSearchRequest(BaseModel):
    query: str
    equipment_type: Optional[str] = None
    n_results: int = 5


class ManualSearchResponse(BaseModel):
    items: list[ManualExcerpt]
    total: int


# ═══════════════════════════════════════════════════════════
# 12. SPC CSV Upload — POST /equipment/spc/upload-csv
# ═══════════════════════════════════════════════════════════


class SPCUploadResponse(BaseModel):
    process_id: str
    n_samples: int
    mean: float
    std: float
    cpk: Optional[float] = None
    grade: str = ""
    violation_count: int = 0


# ═══════════════════════════════════════════════════════════
# 13. Inspection Checklist — GET /equipment/inspection/checklist/{type}
# ═══════════════════════════════════════════════════════════


class ChecklistItem(BaseModel):
    item: str
    standard: str = ""
    unit: str = ""


class ChecklistTemplate(BaseModel):
    id: int
    template_name: str
    equipment_type: str
    checklist_type: str
    items: list[ChecklistItem]


class InspectionChecklistResponse(BaseModel):
    templates: list[ChecklistTemplate]
    total: int

File: 04_AJIN/frontend/src/types/equipment.ts
// Day 6 Phase 2 — 설비/공정 AI (Module F) TS 타입.
// backend/schemas/equipment.py 의 Pydantic 모델과 1:1 정합.

export type Severity = 'critical' | 'warning' | 'info';
export type RiskLevel = 'critical' | 'warning' | 'normal';
export type ProcessStatus = 'good' | 'warning' | 'critical';
export type EngineStatus = 'online' | 'offline' | 'warning';

export type EquipmentSubTab = 'overview' | 'alerts' | 'equipment' | 'predictive' | 'ml';

// 5공정 슬러그 (DAY6_7_PLAN Section 7-1)
export type ProcessSlug = 'cch' | 'obc' | 'bumper_beam' | 'door' | 'ball_seat' | 'cch_plate';

// ── Overview ────────────────────────────────────────────────

export interface ProcessHealthCard {
  process_id: string;
  process_name: string;
  status: ProcessStatus;
  current_cpk: number;
  cpk_trend: string;
  violation_count: number;
  violated_rules: number[];
  risk_level: RiskLevel;
  anomaly_rate: number;
}

export interface EquipmentTypeCard {
  type: string;
  icon: string;
  codes: number;
  key_metric: string;
  color: string;
}

export interface DashboardMetrics {
  error_codes_total: number;
  error_codes_critical: number;
  molds_total: number;
  molds_warning: number;
  molds_critical: number;
  spc_processes: number;
  inspections_templates: number;
  inspections_recent: number;
}

export interface MLAlert {
  level: string;
  source: string;
  message: string;
}

export interface OverviewResponse {
  processes: ProcessHealthCard[];
  equipment_types: EquipmentTypeCard[];
  metrics: DashboardMetrics;
  ml_alerts: MLAlert[];
}

// ── SPC ─────────────────────────────────────────────────────

export interface NelsonViolation {
  rule_number: number;
  rule_name: string;
  description: string;
  severity: Severity;
  points: number[];
  recommended_action: string;
  chart_annotation: string;
}

export interface SPCData {
  process_id: string;
  process_name: string;
  timestamps: number[];
  values: number[];
  mean: number;
  sigma: number;
  ucl: number;
  lcl: number;
  sigma_1_upper: number;
  sigma_1_lower: number;
  sigma_2_upper: number;
  sigma_2_lower: number;
  usl: number | null;
  lsl: number | null;
}

export interface SPCResponse {
  data: SPCData;
  violations: NelsonViolation[];
  out_of_control: boolean;
  violation_count: number;
}

export interface RecentViolation {
  id: string;
  process_id: string;
  process_name: string;
  rule_number: number;
  severity: Severity;
  message: string;
  timestamp: number;
}

export interface ViolationsResponse {
  items: RecentViolation[];
  total: number;
}

// ── Error Search ────────────────────────────────────────────

export interface ErrorSearchResult {
  code: string;
  equipment_type: string;
  category: string;
  description: string;
  cause: string;
  action: string;
  severity: string;
  score: number;
  rank: number;
}

export interface CausalityInfo {
  causes: string[];
  actions: string[];
}

export interface ManualExcerpt {
  content: string;
  source: string;
  page: string;
  relevance: number;
}

export interface ErrorSearchResponse {
  results: ErrorSearchResult[];
  causality: CausalityInfo | null;
  manual_excerpts: ManualExcerpt[];
}

export interface CategoryGroup {
  equipment_type: string;
  symptoms: string[];
}

export interface ErrorCategoriesResponse {
  groups: CategoryGroup[];
  total_synonyms: number;
}

// ── Markov ──────────────────────────────────────────────────

export interface MarkovPrediction {
  code: string;
  category: string;
  equipment_type: string;
  probability: number;
  expected_delay_hours: number;
  description: string;
  recommended_action: string;
}

export interface CascadeStep {
  code: string;
  category: string;
  probability: number;
  expected_delay_hours: number;
}

export interface CascadeChainItem {
  steps: CascadeStep[];
  total_probability: number;
  total_hours: number;
}

export interface MarkovResponse {
  current_code: string;
  current_category: string;
  next_predictions: MarkovPrediction[];
  cascade_chains: CascadeChainItem[];
  risk_level: RiskLevel;
  prevention_message: string;
}

// ── Mold ────────────────────────────────────────────────────

export interface MoldItem {
  mold_id: string;
  mold_name: string;
  mold_type: string;
  part_name: string;
  current_shots: number;
  max_shots: number;
  life_percent: number;
  remaining_shots: number;
  status: string;
  predicted_remaining_life: number | null;
  predicted_replacement_date: string | null;
  risk_level: RiskLevel | null;
  confidence_interval: number[] | null;
}

export interface MoldsResponse {
  items: MoldItem[];
  total: number;
  critical: number;
  warning: number;
  active: number;
}

// ── MTBF ────────────────────────────────────────────────────

export interface MTBFItem {
  machine_id: string;
  machine_name: string;
  total_repairs: number;
  mtbf_days: number;
  mtbf_std_days: number;
  last_repair_date: string;
  next_predicted_date: string;
  days_until_next: number;
  risk_level: string;
  avg_repair_hours: number;
  avg_repair_cost: number;
  seasonal_pattern: Record<string, number>;
}

export interface MTBFTopCost {
  machine_name: string;
  total_cost: number;
}

export interface MTBFResponse {
  items: MTBFItem[];
  top5_cost: MTBFTopCost[];
  seasonal_message: string;
  machines_attention: number;
}

// ── ML Engines ──────────────────────────────────────────────

export interface MLEngineStatus {
  id: string;
  name_en: string;
  name_ko: string;
  library: string;
  status: EngineStatus;
  accuracy: number | null;
  last_trained: string | null;
  description: string;
}

export interface MLEnginesStatusResponse {
  engines: MLEngineStatus[];
  online_count: number;
  total: number;
}

// ── Manual / Inspection ─────────────────────────────────────

export interface ManualSearchResponse {
  items: ManualExcerpt[];
  total: number;
}

export interface ChecklistItem {
  item: string;
  standard: string;
  unit: string;
}

export interface ChecklistTemplate {
  id: number;
  template_name: string;
  equipment_type: string;
  checklist_type: string;
  items: ChecklistItem[];
}

export interface InspectionChecklistResponse {
  templates: ChecklistTemplate[];
  total: number;
}

export interface SPCUploadResponse {
  process_id: string;
  n_samples: number;
  mean: number;
  std: number;
  cpk: number | null;
  grade: string;
  violation_count: number;
}
